enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/log/segmented/SegmentFile.java

```diff
@@ -101,155 +98,120 @@ static SegmentFile create( FileSystemAbstraction fileSystem, File file,
-                    {
-                        // TODO: Consider including position of channel.
-                        readerPool.release( channel );
-                    }
-                };
+        /* This is the relative index within the file, starting from zero. */
+        long offsetIndex = logIndex - (header.prevIndex() + 1);
+
+        LogPosition position = positionCache.lookup( offsetIndex );
+        Reader reader = readerPool.acquire( version, position.byteOffset );
 
-        long currentIndex = position.logIndex;
-        assert currentIndex == 0; // until we properly implement caching
         try
         {
-            while ( currentIndex < offsetIndex )
-            {
-                read( reader, contentMarshal );
-                currentIndex++;
-            }
-        }
-        catch ( ReadPastEndException e )
-        {
-            reader.close();
-            return IOCursor.getEmpty();
-        }
-
-        return new EntryRecordCursor( reader, contentMarshal, currentIndex, logPosition -> {
+            ReadAheadChannel<StoreChannel> bufferedReader = new ReadAheadChannel<>( reader.channel() );
+            long currentIndex = position.logIndex;
             try
             {
-                // TODO: Cache the end position.
-                // TODO: explain why we need the end position in the cache
-                reader.close();
+                /* The cache lookup might have given us an earlier position, scan forward to the exact position. */
+                while ( currentIndex < offsetIndex )
+                {
+                    read( bufferedReader, contentMarshal );
+                    currentIndex++;
+                }
             }
-            catch ( IOException e )
+            catch ( ReadPastEndException e )
             {
-                log.error( "Failed to close reader: " + file );
+                bufferedReader.close();
+                return IOCursor.getEmpty();
             }
-        } );
-    }
 
-    private LogPosition findCachedStartingPosition( long offsetIndex )
-    {
-        // TODO: Implement cache lookups.
-        return new LogPosition( 0, SegmentHeader.SIZE );
-    }
+            return new EntryRecordCursor( bufferedReader, contentMarshal, currentIndex )
+            {
+                boolean closed = false; /* This is just a defensive measure, for catching user errors from messing up the refCount. */
 
-    private PhysicalFlushableChannel getOrCreateWriter() throws IOException
-    {
-        if ( closed )
+                @Override
+                public void close()
+                {
+                    if ( closed )
+                    {
+                        throw new IllegalStateException( "Already closed" );
+                    }
+
+                    closed = true;
+                    positionCache.put( this.position() );
+                    readerPool.release( reader );
+                    refCount.decrease();
```

Comment:
So we specifically do not want to close it, thus no super.close(), but I will take a look at extracting this and make it clearer.